File: 902_Kth_Smallest_Element_in_a_BST.py
#!/usr/bin/python3

'''
902. Kth Smallest Element in a BST
Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

You may assume k is always valid, 1 ≤ k ≤ BST's total elements.

Example
Given root = {1,#,2}, k = 2, return 2.

Challenge
What if the BST is modified (insert/delete operations) often and you need to find the kth smallest frequently? How would you optimize the kthSmallest routine?


Author: Yuan Li
Date: 6/27/2018 
Difficulty: Medium
'''

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: the given BST
    @param k: the given k
    @return: the kth smallest element in BST
    """
    def __init__(self):
        self.count = 0
        self.res = None
    
    def kthSmallest(self, root, k):
        # write your code here
        self.inOrder(root, k)
        return self.res
        
    def inOrder(self, root, k):
        if root:
            self.inOrder(root.left, k)

            self.count += 1
            if self.count == k:
                self.res = root.val
                return
            
            self.inOrder(root.right, k)
        

File: 902_Kth_Smallest_Element_in_a_BST.cpp
/*
902. Kth Smallest Element in a BST
Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

Example
Given root = {1,#,2}, k = 2, return 2.

Challenge
What if the BST is modified (insert/delete operations) often and you need to find the kth smallest frequently? How would you optimize the kthSmallest routine?

Notice
You may assume k is always valid, 1 ≤ k ≤ BST's total elements.


Author: Yuan Li
Date: 9/6/2018 
Difficulty: Medium
*/

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class Solution {
  public:
    /**
     * @param root: the given BST
     * @param k: the given k
     * @return: the kth smallest element in BST
     */
    int kthSmallest(TreeNode * root, int k) {
        // write your code here
        int res;
        int cnt = 0;
        
        while (root) {
            stk_.push(root);
            root = root->left;
        }
        
        while (!stk_.empty()) {
            TreeNode* node = stk_.top();
            ++cnt;
            if (cnt == k) {
                res = node->val;
                break;
            }
            
            if (node->right) { // look for the left-most of right subtree
                node = node->right;
                while (node) {
                    stk_.push(node);
                    node = node->left;
                }
            } else { // look for the first left turn
                stk_.pop();
                while (!stk_.empty() && stk_.top()->right == node) {
                    node = stk_.top();
                    stk_.pop();
                }
            }
        }
        
        return res;
    }
    
  private:
    stack<TreeNode*> stk_;
};
